File: 0072-edit-distance/0072-edit-distance.java
class Solution {
    public int dfs(int i, int j, String s, String t, int[][] dp){
        if(i == s.length() && j == t.length()) return 0;

        if(i == s.length()) return t.length() - j; 

        if(j == t.length()) return s.length() - i; 

        if(dp[i][j] != -1) return dp[i][j];

        int res = Integer.MAX_VALUE;
        if(s.charAt(i) == t.charAt(j)){
            res = dfs(i+1, j+1, s, t, dp);
        }
        else{
            res = 1 + Math.min(dfs(i+1, j+1, s, t, dp), 
                    Math.min(dfs(i, j+1, s, t, dp), dfs(i+1, j, s, t, dp)));
        }

        dp[i][j] = res;
        return res;
    }

    public int minDistance(String word1, String word2) {
        int[][] dp = new int[word1.length()+1][word2.length()+1];

        for(int i=0; i<=word1.length(); i++){
            for(int j=0; j<=word2.length(); j++){
                dp[i][j] = -1;
            }
        }

        return dfs(0, 0, word1, word2, dp);
    }
}

File: 0072-edit-distance/0072-edit-distance.py
class Solution(object):
    def minDistance(self, text1, text2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """

        m = len(text1)
        n = len(text2)

        dp = [[0 for _ in range(n+1)] for _ in range(m+1)]

        dp[0][n] = len(text1)
        for i in range(1, m):
            dp[i][n] = dp[i-1][n] - 1

        dp[m][0] = len(text2)
        for i in range(1, n):
            dp[m][i] = dp[m][i-1] - 1


        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                if text1[i] == text2[j]:
                    dp[i][j] = dp[i+1][j+1]

                else:
                    dp[i][j] = min(dp[i+1][j], dp[i][j+1], dp[i+1][j+1]) + 1

        return dp[0][0]
        
        